File: keep_logon.py
from datetime import datetime, timezone
from Crypto.Signature import eddsa
from Crypto.PublicKey import ECC

import os
import base64
import socket
import ssl
import random
import string


def get_time():
    return datetime.now(timezone.utc).strftime('%Y%m%d-%H:%M:%S.%f')[:-3]

def sign(payload):
    sign = eddsa.new(private_key,'rfc8032')
    sign_data = sign.sign(payload.encode("ASCII"))                        # 签名
    signature = base64.b64encode(sign_data).decode('ASCII')  
    return signature

def get_final_message(fix_message):
    
    temp_str = f"8=FIX.4.4|9={len(fix_message)}|" + fix_message
    chesum = sum(ord(char) for char in temp_str.replace('|','\x01'))
    final_msg = temp_str + f"10={chesum%256:03}|"

    return final_msg.replace('|','\x01')



api_key = ""

dir = os.path.dirname(__file__)
f = open(dir+"/private_key.pem", "rb")
private_key = f.read()                                # 获取私钥
private_key = ECC.import_key(private_key)



utc_timestamp = get_time()
sender_comp_id=''.join(random.sample(string.ascii_letters + string.digits,8))
target_comp_id='SPOT'
msg_seq_num=1
payload = chr(1).join(['A',sender_comp_id,target_comp_id,str(msg_seq_num),utc_timestamp])
signature = sign(payload)

fix_message=(f"35=A|"f"49={sender_comp_id}|"f"56={target_comp_id}|"f"34=1|"f"52={utc_timestamp}|"
f"95={len(signature)}|"f"96={signature}|"f"98=0|"f"108=10|"f"141=Y|"f"553={api_key}|"f"25035=2|")


sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
ssl_sock = ssl.wrap_socket(sock)
ssl_sock.connect(('fix-oe.testnet.binance.vision', 9000))

send_msg = get_final_message(fix_message)
print(type(send_msg.encode('ASCII')))
print("send: ",send_msg.replace('\x01','|'))
# send logon message
ssl_sock.sendall(send_msg.encode('ASCII'))
msg_seq_num += 1
# get response


while True:

    response = ssl_sock.recv(4096)

    if response:
        
        data_str = response.decode('ASCII')
        print("recv: ",data_str.replace("\x01","|"))
        data_list = data_str.split('\x01')

        if data_list[2] == '35=1':  # send Heartbeat to keep connected
            TestReqID = data_list[-3]
            #print(TestReqID)
            fix_message=(f"35=0|"f"49={sender_comp_id}|"f"56={target_comp_id}|"f"34={msg_seq_num}|"f"52={get_time()}|"f"{TestReqID}|")
            send_msg = get_final_message(fix_message)
            print("send: ",send_msg.replace('\x01','|'))
            ssl_sock.sendall(send_msg.encode('ASCII'))
            msg_seq_num += 1
            

File: keep_logon.js
const crypto = require("crypto")
const fs = require("fs")
const tls = require("tls")



let utc_timestamp=get_time()
let sender_comp_id=random_string()
let target_comp_id='SPOT'
let msg_seq_num=1
let payload = ['A',sender_comp_id,target_comp_id,msg_seq_num,utc_timestamp].join(String.fromCharCode(1))
//console.log(payload)


let api_key = ""
let private_key = fs.readFileSync(__dirname+'/private_key.pem');
let private_key_obj = crypto.createPrivateKey(private_key);
let signature = crypto.sign(null,payload,private_key_obj).toString('base64')

//console.log(signature)
let fix_message = `35=A|49=${sender_comp_id}|56=${target_comp_id}|34=1|52=${utc_timestamp}|\
95=${signature.length}|96=${signature}|98=0|108=60|141=Y|553=${api_key}|25035=2|`

//console.log(fix_message)


let final_msg = get_final_message(fix_message)




const client = tls.connect(9000,"fix-oe.binance.com",
    ()=>{
        
        client.write(final_msg,'ascii');
        console.log("send: "+final_msg.replaceAll(String.fromCharCode(1),'|'))
        msg_seq_num++;
    });

    client.once('connect',()=>{
        
    })

    client.on('data',(data)=>{
        
        let response = data.toString();
        console.log("recv: "+response.replaceAll(String.fromCharCode(1),'|'));

        let data_list = response.split(String.fromCharCode(1));
        if(data_list[2]==='35=1'){
            let hearbeat_msg = `35=0|49=${sender_comp_id}|56=${target_comp_id}|34=${msg_seq_num}|52=${get_time()}|${data_list.at(-3)}|`;
            final_msg = get_final_message(hearbeat_msg);
            client.write(final_msg,'ascii');
            console.log("send: "+final_msg.replaceAll(String.fromCharCode(1),'|'))
            msg_seq_num++;
        }
    })

    client.on('error',(error)=>{
        console.log(error)
    });

    client.on('end',()=>{
        console.log('disconnected')
    });


    process.stdin.resume();







function get_final_message(fix_message){
    let temp_str = `8=FIX.4.4|9=${fix_message.length}|` + fix_message
    checksum = temp_str.replaceAll('|',String.fromCharCode(1)).split('').map(element=> element.charCodeAt()).reduce(
        (accumulator, currentValue) => accumulator + currentValue
      )
    let final_msg = temp_str + `10=${String(checksum%256).padStart(3,'0')}|`

    return final_msg.replaceAll('|',String.fromCharCode(1))
}
    




function get_time(){
    time = new Date()
    return time.getUTCFullYear()+
    String(time.getUTCMonth()+1).padStart(2, '0')+
    String(time.getUTCDate()).padStart(2,'0')+
    '-'+
    String(time.getUTCHours()).padStart(2,'0')+':'+
    String(time.getUTCMinutes()).padStart(2,'0')+':'+
    String(time.getUTCSeconds()).padStart(2,'0')+
    '.'+
    String(time.getMilliseconds()).padStart(3,'0')
}


function random_string(k){
    k = k || 8;
    let characters = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
    let c ='';
    for(let i=0; i<k; i++){
        c += characters.charAt(Math.floor(Math.random()*characters.length))
    }
    return c;
}
